### PR TITLE
raft: leader respond to learner read index message

### LINKED ISSUE #10589: Leader does not respond to ReadIndex message from learner
In a cluster with 1 leader and 1 learner (quorum == 1), a linearizable read request sent to learner node will timeout. More detailed logs can be found [in this issue](https://github.com/jingyih/etcd/issues/7). Basically, the learner node will timeout after sending out ReadIndex message to leader and waiting for leader's response.

**What is supposed to happen**

1. A linearizable read request notifies the `LinearizableReadLoop`.
https://github.com/etcd-io/etcd/blob/9c2b88d783411d2e100385ca5ded21cd5796167c/etcdserver/v3_server.go#L94-L96

2. Upon being notified, `LinearizableReadLoop` sends a ReadIndex request to local raft node. In this case, this node is a raft learner.
https://github.com/etcd-io/etcd/blob/9c2b88d783411d2e100385ca5ded21cd5796167c/etcdserver/v3_server.go#L651

3. The learner node, also as a follower, upon receiving ReadIndex message, forwards the message to leader.
https://github.com/etcd-io/etcd/blob/9c2b88d783411d2e100385ca5ded21cd5796167c/raft/raft.go#L1291-L1297

4. Leader receives the message, sends out heartbeat to ensure it is not deposed by a more recent leader (for simplicity let's only talk about `ReadOnlySafe` option)
https://github.com/etcd-io/etcd/blob/9c2b88d783411d2e100385ca5ded21cd5796167c/raft/raft.go#L1025-L1039

5. Eventually leader sends out ReadIndex response after all safety checks are passed.
https://github.com/etcd-io/etcd/blob/9c2b88d783411d2e100385ca5ded21cd5796167c/raft/raft.go#L1142

**What went wrong**
In a cluster with 1 leader and 1 learner, raft quorum is 1. In step 4, leader automatically assumes the ReadIndex request is from local, appends ReadIndex information to local readState, and return. Therefore in this case, the learner will not get a response from leader, which leads to request timeout.
https://github.com/etcd-io/etcd/blob/9c2b88d783411d2e100385ca5ded21cd5796167c/raft/raft.go#L1047-L1049

### PR DESCRIPTION
**Pull request by jingyih**
Leader should check message sender upon receiving MsgReadIndex message, even when raft quorum is 1. The message could be sent from learner node. In this case, leader should respond to the sender.

Fixes #10589.

cc @xiang90 @gyuho @WIZARD-CXY @jpbetz

### CODE REVIEW

**Comment by jingyih** [comment]
In this case, does the leader need to check if it has committed any log entry at the current term?

File: raft/raft.go [hunk 1]
```diff
@@ -1045,7 +1045,11 @@ func stepLeader(r *raft, m pb.Message) error {
 				}
 			}
 		} else {
-			r.readStates = append(r.readStates, ReadState{Index: r.raftLog.committed, RequestCtx: m.Entries[0].Data})
+			if m.From == None || m.From == r.id { // from local member
+				r.readStates = append(r.readStates, ReadState{Index: r.raftLog.committed, RequestCtx: m.Entries[0].Data})
+			} else {
```
Comment (xiang90) [inline raft/raft.go]: can you add comment on this case?

**Comment by xiang90** [comment]
lgtm

File: raft/raft.go (hunk 1 above)
Comment (jingyih) [reply raft/raft.go]: Done. PTAL.

File: raft/raft.go [hunk 2]
```diff
@@ -1044,8 +1044,12 @@ func stepLeader(r *raft, m pb.Message) error {
 					r.send(pb.Message{To: m.From, Type: pb.MsgReadIndexResp, Index: ri, Entries: m.Entries})
 				}
 			}
-		} else {
-			r.readStates = append(r.readStates, ReadState{Index: r.raftLog.committed, RequestCtx: m.Entries[0].Data})
+		} else { // there is only one voting member (the leader) in the cluster
+			if m.From == None || m.From == r.id { // from leader itself
+				r.readStates = append(r.readStates, ReadState{Index: r.raftLog.committed, RequestCtx: m.Entries[0].Data})
+			} else { // from learner member
```
Comment (xiang90) [inline raft/raft.go]: i just hit the merge button... but actually you can check if m.From is indeed a learner for extra safety.

Comment (jingyih) [reply raft/raft.go]: Since here we have `r.quorum()==1`, which means `r.prs` only has leader itself?

**Pull request merged**